analyses/cms-open-data-ttbar/analysis.py

```diff
@@ -122,80 +122,85 @@ def make_rdf(
-    df = df.Filter("Sum(jet_btag[jet_pt_mask]>=0.5)>1")
+    df = df.Filter("Sum(Jet_btagCSVV2[Jet_pt_mask]>0.5)>1")
 
     # Build four-momentum vectors for each jet
-    df = df.Define(
-        "jet_p4",
+    df = (  
+        df.Define(
+        "Jet_p4",
         """
         ROOT::VecOps::Construct<ROOT::Math::PxPyPzMVector>(
-            jet_px[jet_pt_mask], jet_py[jet_pt_mask], jet_pz[jet_pt_mask], jet_mass[jet_pt_mask])
+            ROOT::VecOps::Construct<ROOT::Math::PtEtaPhiMVector>(
+                Jet_pt[Jet_pt_mask], Jet_eta[Jet_pt_mask], Jet_phi[Jet_pt_mask], Jet_mass[Jet_pt_mask]
+            )
+        )
         """,
+        )
     )
 
     # Build trijet combinations
-    df = df.Define("trijet", "ROOT::VecOps::Combinations(jet_pt[jet_pt_mask],3)")
-    df = df.Define("ntrijet", "trijet[0].size()")
+    df = df.Define("Trijet", "ROOT::VecOps::Combinations(Jet_pt[Jet_pt_mask],3)")
+    df = df.Define("nTrijet", "Trijet[0].size()")
 
     # Assign four-momentums to each trijet combination
     df = df.Define(
-        "trijet_p4",
+        "Trijet_p4",
         """
-        ROOT::RVec<ROOT::Math::PxPyPzMVector> trijet_p4(ntrijet);
-        for (int i = 0; i < ntrijet; ++i)
+        ROOT::RVec<ROOT::Math::PxPyPzMVector> Trijet_p4(nTrijet);
+        for (int i = 0; i < nTrijet; ++i)
         {
-            int j1 = trijet[0][i];
-            int j2 = trijet[1][i];
-            int j3 = trijet[2][i];
-            trijet_p4[i] = jet_p4[j1] + jet_p4[j2] + jet_p4[j3];
+            int j1 = Trijet[0][i];
+            int j2 = Trijet[1][i];
+            int j3 = Trijet[2][i];
+            Trijet_p4[i] = Jet_p4[j1] + Jet_p4[j2] + Jet_p4[j3];
         }
-        return trijet_p4;
+        return Trijet_p4;
         """,
     )
 
     # Get trijet transverse momentum values from four-momentum vectors
     df = df.Define(
-        "trijet_pt",
-        "return ROOT::VecOps::Map(trijet_p4, [](ROOT::Math::PxPyPzMVector v) { return v.Pt(); })",
+        "Trijet_pt",
+        "return ROOT::VecOps::Map(Trijet_p4, [](const ROOT::Math::PxPyPzMVector &v) { return v.Pt(); })",
     )
 
     # trijet_btag is a helpful array of bool values indicating whether or not the maximum btag value in trijet is larger than 0.5 threshold
     df = df.Define(
-        "trijet_btag",
+        "Trijet_btag",
         """
-        ROOT::RVecB btag(ntrijet);
-        for (int i = 0; i < ntrijet; ++i)
+        ROOT::RVecB btag(nTrijet);
+        for (int i = 0; i < nTrijet; ++i)
         {
-            int j1 = trijet[0][i];
-            int j2 = trijet[1][i];
-            int j3 = trijet[2][i];
-            btag[i] = std::max({jet_btag[j1], jet_btag[j2], jet_btag[j3]}) > 0.5;
+            int j1 = Trijet[0][i];
+            int j2 = Trijet[1][i];
+            int j3 = Trijet[2][i];
+            btag[i] = std::max({Jet_btagCSVV2[Jet_pt_mask][j1], Jet_btagCSVV2[Jet_pt_mask][j2], Jet_btagCSVV2[Jet_pt_mask][j3]}) > 0.5;
```

Comment:
we are applying the same mask to the same array 3 times here. it's a bit wasteful.

in fact, every use of `Jet_btagCSSV2` comes with the `Jet_pt_mask` attached if I'm not mistaken, so we can just `Define("Jet_btag_masked", ...)` once and then use that.

Reply:
This is resolved in the reimplemented algorithm.

Reply:
ok!